ChildProject/pipelines/metrics.py

```diff
@@ -80,46 +120,262 @@ def __init__(
             self.child_cols = None
         
         self.by = by
+        self.period = period
+        
+        #build a dataframe with all the periods we will want for each unit
+        if self.period: 
+            self.periods = pd.interval_range(
+                start=datetime.datetime(1900, 1, 1, 0, 0, 0, 0),
+                end=datetime.datetime(1900, 1, 2, 0, 0, 0, 0),
+                freq=self.period,
+                closed="both",
+            )
+            self.periods= pd.DataFrame(self.periods.to_tuples().to_list(),columns=['period_start','period_end'])
+        
         self.segments = pd.DataFrame()
 
         self.recordings = Pipeline.recordings_from_list(recordings)
-
-        self.from_time = from_time
-        self.to_time = to_time
+        
+        if from_time:
+            try:
+                self.from_time = datetime.datetime.strptime(from_time, "%H:%M")
+            except:
+                raise ValueError(f"invalid value for from_time ('{from_time}'); should have HH:MM format instead")
+        else:
+            self.from_time = None
+        
+        if to_time:
+            try:
+                self.to_time = datetime.datetime.strptime(to_time, "%H:%M")
+            except:
+                raise ValueError(f"invalid value for to_time ('{to_time}'); should have HH:MM format instead")
+        else:
+            self.to_time = None
+        
+        self._initiate_metrics_df()
 
     def __init_subclass__(cls, **kwargs):
         super().__init_subclass__(**kwargs)
         pipelines[cls.SUBCOMMAND] = cls
 
-    @abstractmethod
+    def _process_unit(self,row):
+        """for one unit (i.e. 1 {recording|session|child} [period]) compute the list of required metrics and store the results in the current row of self.metrics
+        
+        :param row: index and Series of the unit to process, to be modified with the results
+        :type row: (int , pandas.Series)
+        :return: Series containing all the computed metrics result for that unit
+        :rtype: pandas.Series
+        """
+        prev_set = ""
+        duration_set = 0
+        for i, line in self.metrics_list.iterrows():
+            curr_set = line["set"]
+            if prev_set != curr_set:
+                index, annotations = self.retrieve_segments([curr_set],row[1])
+                if index.shape[0]:
+                    duration_set = (
+                            index["range_offset"] - index["range_onset"]
+                        ).sum()
+                else : duration_set = 0
+                row[1]["duration_{}".format(line["set"])] = duration_set
```

Comment:
Could you add a comment to explain what row[1] is? (and what is row[0] ?)

Reply:
7daaba6